eclair-core/src/main/scala/fr/acinq/eclair/payment/Bolt12Invoice.scala

```diff
@@ -113,72 +104,75 @@ case class Bolt12Invoice(records: TlvStream[InvoiceTlv]) extends Invoice {
       offerId.contains(offer.offerId) &&
       offer.chains.contains(chain) &&
+      request.chain == chain &&
       !isExpired() &&
-      request.amount.contains(amount) &&
+      // The recipient is allowed to provide a discount (e.g. when paying for multiple items).
+      request.amount.forall(requestedAmount => amount <= requestedAmount) &&
       quantity == request.quantity_opt &&
       payerKey.contains(request.payerKey) &&
       payerInfo == request.payerInfo &&
-      payerNote.forall(request.payerNote.contains(_)) && // It's OK to have a payer's note in the request but not in the invoice.
-      description == Left(offer.description) &&
+      // Bolt 12: MUST reject the invoice if payer_note is set, and was unset or not equal to the field in the invoice_request.
+      payerNote.forall(request.payerNote.contains(_)) &&
       issuer == offer.issuer &&
       request.features.areSupported(features)
   }
 
   def checkRefundSignature(): Boolean = {
     (refundSignature, refundFor, payerKey) match {
-      case (Some(sig), Some(hash), Some(key)) =>
-        verifySchnorr("lightning" + "invoice" + "payer_signature", hash, sig, key)
+      case (Some(sig), Some(hash), Some(key)) => verifySchnorr(signatureTag("payer_signature"), hash, sig, key)
       case _ => false
     }
   }
 
   def checkSignature(): Boolean = {
-    val withoutSig = TlvStream(records.records.filter { case _: Signature => false case _ => true }, records.unknown)
-    verifySchnorr("lightning" + "invoice" + "signature", rootHash(withoutSig, invoiceTlvCodec).get, signature, ByteVector32(nodeId.value.drop(1)))
+    verifySchnorr(signatureTag("signature"), rootHash(Offers.removeSignature(records), invoiceTlvCodec), signature, Offers.xOnlyPublicKey(nodeId))
   }
-
-  def withNodeId(id: PublicKey): Bolt12Invoice =
-    Bolt12Invoice(TlvStream(records.records.map { case NodeId(_) => NodeId(id) case x => x }, records.unknown))
 }
 
 object Bolt12Invoice {
+  val hrp = "lni"
   val DEFAULT_EXPIRY_SECONDS: Long = 7200
+  val DEFAULT_MIN_FINAL_EXPIRY_DELTA: CltvExpiryDelta = CltvExpiryDelta(18)
 
-  case class BlindedPath(route: RouteBlinding.BlindedRoute, payInfos: Seq[PayInfo], capacity: Option[MilliSatoshi])
+  case class BlindedPathWithPayInfo(route: RouteBlinding.BlindedRoute, payInfo: PayInfo, capacity: Option[MilliSatoshi])
 
   /**
-   * Creates an invoice for a given offer and invoice request
-   * @param offer    the offer this invoices corresponds to
-   * @param request  the request this invoice responds to
-   * @param preimage the preimage to use for the payment
-   * @param nodeKey  the key that was used to generate the offer, may be different from our public nodeId if we're hiding behind a blinded route
-   * @param features invoice features
+   * Creates an invoice for a given offer and invoice request.
+   *
+   * @param amount    the amount of the invoice (which may be different from the offer and invoice request, for example when applying a discount)
```

Comment:
That should be clarified in the offers spec PR, because this requirement indicates that it's allowed to *not* use the `invoice_request` amount:

```
- if it does not set amount to the base invoice amount calculated from the invoice_request:
    - MUST append the reason to description (e.g. " 5% bulk discount").
```

Maybe we just don't want to support this scenario at all though? I agree that the UX around this would be hard to get right.